nomad/encrypter.go

```diff
@@ -341,21 +344,147 @@ func (e *Encrypter) VerifyClaim(tokenString string) (*structs.IdentityClaims, er
 	return claims, nil
 }
 
-// AddKey stores the key in the keystore and creates a new cipher for it.
-func (e *Encrypter) AddKey(rootKey *structs.RootKey) error {
+// AddUnwrappedKey stores the key in the keystore and creates a new cipher for
+// it. This is called in the RPC handlers on the leader and from the legacy
+// KeyringReplicator.
+func (e *Encrypter) AddUnwrappedKey(rootKey *structs.RootKey, isUpgraded bool) (*structs.WrappedRootKeys, error) {
 
 	// note: we don't lock the keyring here but inside addCipher
 	// instead, so that we're not holding the lock while performing
 	// local disk writes
 	if err := e.addCipher(rootKey); err != nil {
-		return err
+		return nil, err
 	}
-	if err := e.saveKeyToStore(rootKey); err != nil {
-		return err
+	return e.wrapRootKey(rootKey, isUpgraded)
+}
+
+// AddWrappedKey creates decryption tasks for keys we've previously stored in
+// Raft. It's only called as a goroutine by the FSM Apply for WrappedRootKeys,
+// but it returns an error for ease of testing.
+func (e *Encrypter) AddWrappedKey(ctx context.Context, wrappedKeys *structs.WrappedRootKeys) error {
+
+	logger := e.log.With("key_id", wrappedKeys.KeyID)
+
+	e.lock.Lock()
+
+	_, err := e.keysetByIDLocked(wrappedKeys.KeyID)
+	if err == nil {
+
+		// key material for each key ID is immutable so nothing to do, but we
+		// can cancel and remove any running decrypt tasks
+		if cancel, ok := e.decryptTasks[wrappedKeys.KeyID]; ok {
+			cancel()
+			delete(e.decryptTasks, wrappedKeys.KeyID)
+		}
+		e.lock.Unlock()
+		return nil
+	}
+
+	if cancel, ok := e.decryptTasks[wrappedKeys.KeyID]; ok {
+		// stop any previous tasks for this same key ID under the assumption
+		// they're broken or being superceded, but don't remove the CancelFunc
+		// from the map yet so that other callers don't think we can continue
+		cancel()
+	}
+
+	e.lock.Unlock()
+
+	completeCtx, cancel := context.WithCancel(ctx)
+
+	for _, wrappedKey := range wrappedKeys.WrappedKeys {
+		providerID := wrappedKey.ProviderID
+		if providerID == "" {
+			providerID = string(structs.KEKProviderAEAD)
+		}
+
+		provider, ok := e.providerConfigs[providerID]
+		if !ok {
+			logger.Error("no such KMS provider configured - root key cannot be decrypted",
+				"provider_id", providerID)
+			cancel()
+			return fmt.Errorf("no such provider %q configured", providerID)
```

Comment:
I think we can do that. We'll probably want to capture whether _any_ of the providers managed to get to `go e.decryptWrappedKeyTask` and if not, call `cancel` so that we're not leaving the context's goroutine sitting around doing nothing. We'll still have the function in the map so that'll block `IsReady` and force the user to fix their config.